### PR TITLE
enh(api): add endpoint to find access groups

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to add endpoint find access groups use case.

For a user admin return all the access groups.
image

For a user submitted to ACL, return the access groups where he is member of.
image

**Fixes** # MON-13478

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Security/Application/UseCase/FindUserAccessGroups/FindUserAccessGroupsResponse.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * @param array<AccessGroup> $accessGroups
+     * @return array<array<string,string|int>>
+     */
+    private function accessGroupsToArray(array $accessGroups): array
+    {
+        return array_map(
+            fn (AccessGroup $accessGroup) => [
```
Comment (callapa) [inline src/Core/Security/Application/UseCase/FindUserAccessGroups/FindUserAccessGroupsResponse.php:50]: Why not include the other fields (alias, activated, group_changed) ?

File: src/Core/Security/Infrastructure/Repository/DbReadAccessGroupRepository.php [hunk 2]
```diff
@@ -1,65 +1,121 @@
...
+        $sortRequest = $this->sqlRequestTranslator->translateSortParameterToSql();
+        $request .= $sortRequest !== null ? $sortRequest : ' ORDER BY acl_group_id ASC';
+
+        // Pagination
+        $request .= $this->sqlRequestTranslator->translatePaginationToSql();
+
+        $statement = $this->db->prepare($request);
+
+        foreach ($this->sqlRequestTranslator->getSearchValues() as $key => $data) {
+            $type = key($data);
```
Comment (TamazC) [inline src/Core/Security/Infrastructure/Repository/DbReadAccessGroupRepository.php:86]: are you sure about using of key function into the loop ?
https://www.php.net/manual/fr/function.key.php#114905

File: src/Core/Security/Infrastructure/Repository/DbReadAccessGroupRepository.php [hunk 3]
```diff
@@ -75,18 +131,77 @@ public function findByContact(ContactInterface $contact): array
...
+            $sortRequest = $this->sqlRequestTranslator->translateSortParameterToSql();
+            $request .= $sortRequest !== null ? $sortRequest : ' ORDER BY acl_group_id ASC';
+
+            // Pagination
+            $request .= $this->sqlRequestTranslator->translatePaginationToSql();
+
+            $statement = $this->db->prepare($request);
+
+            foreach ($this->sqlRequestTranslator->getSearchValues() as $key => $data) {
+                $type = key($data);
```
Comment (TamazC) [inline src/Core/Security/Infrastructure/Repository/DbReadAccessGroupRepository.php:186]: same here

File: src/Core/Security/Infrastructure/Api/FindUserAccessGroups/FindUserAccessGroupsController.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+     * @return object
+     */
+    public function __invoke(FindUserAccessGroups $useCase, FindUserAccessGroupsPresenterInterface $presenter): object
+    {
+        $this->denyAccessUnlessGrantedForApiConfiguration();
+
+        /**
+         * @var Contact
+         */
+        $request = new FindUserAccessGroupsRequest($this->getUser());
```
Comment (callapa) [inline src/Core/Security/Infrastructure/Api/FindUserAccessGroups/FindUserAccessGroupsController.php:48]: You can't put an object in a DTO, only scalar values are allowed.

**Review by callapa — CHANGES_REQUESTED** [review]

File: src/Core/Security/Application/UseCase/FindUserAccessGroups/FindUserAccessGroupsRequest.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+     * @param int $userId
+     */
+    public function __construct(public int $userId)
+    {
+    }
+
+    /**
+     * @var bool
+     */
+    public bool $isUserAdmin = false;
```
Comment (kduret) [inline src/Core/Security/Application/UseCase/FindUserAccessGroups/FindUserAccessGroupsRequest.php:38]: move property before construct

File: src/Core/Security/Application/Repository/ReadAccessGroupRepositoryInterface.php [hunk 6]
```diff
@@ -28,11 +28,26 @@
...
      */
     public function findByContact(ContactInterface $contact): array;
+
+    /**
+     * Find all access groups according to a contact with filter
+     *
+     * @param int $userId
+     * @return AccessGroup[]
+     */
+    public function findByContactWithFilter(int $userId): array;
```
Comment (kduret) [inline src/Core/Security/Application/Repository/ReadAccessGroupRepositoryInterface.php:52]: `WithFilter` means using query search parameter ?
if yes, it's not consistent with `findAll` which also use query search parameter

**Review by callapa — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**